recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb

```diff
@@ -0,0 +1,810 @@
+    "\n",
+    "parameters = {\n",
+    "    GenParams.DECODING_METHOD: DecodingMethods.GREEDY,\n",
+    "    GenParams.MIN_NEW_TOKENS: 1,\n",
+    "    GenParams.MAX_NEW_TOKENS: 100,\n",
+    "    GenParams.STOP_SEQUENCES: [\"<|endoftext|>\"]\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Initialize the WatsonX model\n",
+    "Using a CustomLLM wrapper from Langchain, initialize the model with defined parameters and the model specified above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 40,
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "from langchain_ibm import WatsonxLLM\n",
+    "\n",
+    "watsonx_granite = WatsonxLLM(\n",
+    "    model_id=model_id.value,\n",
+    "    url=credentials.get(\"url\"),\n",
+    "    apikey=credentials.get(\"apikey\"),\n",
+    "    project_id=project_id,\n",
+    "    params=parameters\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Answering Queries"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Automate the RAG pipeline\n",
+    "\n",
+    "Build a question-answering chain with the model and the document retriever."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 41,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.chains import RetrievalQA\n",
+    "\n",
+    "qa = RetrievalQA.from_chain_type(llm=watsonx_granite, chain_type=\"stuff\", retriever=docsearch.as_retriever())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "### Generate a retrieval-augmented response to a question"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "Use the question-answering chain to process the query."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'query': 'What did the president say about Ketanji Brown Jackson',\n",
+       " 'result': ' The president said, \"One of our nation’s top legal minds, who will continue Justice Breyer’s legacy of excellence.\"'}"
+      ]
+     },
+     "execution_count": 42,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "query = \"What did the president say about Ketanji Brown Jackson\"\n",
```

Comment:
It took several minutes for an answer to come back! I don't know if it was something in my setup, but you might want to add a warning that sometimes it takes a long time to return an answer (in this "lightweight test setup"?).

Reply:
It was quick for me. Maybe longer for you since you are using the "Lite" plan? I'll put a note in.

Reply:
Does it come back faster the second time?